crates/codegen/grammar/src/dsl.rs

```diff
@@ -0,0 +1,509 @@
+            $name
+            $value
+        }
+    };
+    ($context:ident precedence parser $name:ident $value:tt) => {
+        slang_grammar_definition! {
+            $context
+            PrecedenceParserDefinition
+            PrecedenceParserDefinitionRef
+            PrecedenceParserDefinitionNode
+            slang_precedence_parser
+            None
+            $name
+            $value
+        }
+    };
+}
+
+#[macro_export]
+macro_rules! slang_grammar_definition {
+    ($context:ident $trait:ident $trait_ref:ident $node_type:ident $dsl_macro:ident $is_inline:tt $name:ident $value:tt) => {
+        #[derive(Debug)]
+        struct $name {
+            node: once_cell::unsync::OnceCell<$crate::$node_type>,
+        }
+
+        impl $name {
+            const SOURCE_LOCATION: $crate::SourceLocation = slang_location!();
+            const NAME: &str = stringify!($name);
+            const INSTANCE: once_cell::unsync::OnceCell<std::rc::Rc<Self>> =
+                once_cell::unsync::OnceCell::new();
+            fn instance() -> $crate::$trait_ref {
+                Self::INSTANCE
+                    .get_or_init(|| {
+                        std::rc::Rc::new(Self {
+                            node: once_cell::unsync::OnceCell::new(),
+                        })
+                    })
+                    .clone()
+            }
+        }
+
+        impl $crate::$trait for $name {
+            fn name(&self) -> &'static str {
+                Self::NAME
+            }
+            fn source_location(&self) -> $crate::SourceLocation {
+                Self::SOURCE_LOCATION
+            }
+            fn node(&self) -> &$crate::$node_type {
+                &self.node.get_or_init(|| $dsl_macro!($value))
+            }
+            slang_grammar_definition! { @context_method $context }
+            slang_grammar_definition! { @is_inline_method $is_inline }
+        }
+    };
+
+    (@context_method None ) => {};
+    (@context_method $context:ident ) => {
+        fn context(&self) -> &'static str {
+            stringify!($context)
+        }
+    };
+
+    (@is_inline_method None ) => {};
+    (@is_inline_method ($is_inline:literal) ) => {
+        fn is_inline(&self) -> bool {
+            $is_inline
+        }
+    };
+}
+
+#[macro_export]
+macro_rules! slang_grammar_register_definitions {
+
+    ($grammar:ident lexical context $name:ident { $( $($idents:ident)+ = $value:tt ; )* } ) => {
+        $(
+            slang_grammar_register_definitions! { $grammar $($idents)+ $value }
+        )*
+    };
+    ($grammar:ident config $value:tt) => {};
+    ($grammar:ident $(inline)? $(trivia)? $(precedence)? parser $name:ident $value:tt) => {
+        $grammar.register($name::instance());
+    };
+    ($grammar:ident scanner $name:ident $value:tt) => {
+        $grammar.register($name::instance());
+    };
+
+}
+
+#[macro_export]
+macro_rules! slang_location {
+    // Meaningless in macro_rules! because the locations are all wrong.
```

Comment:
IIRC, we are moving to procedural macros to get source locations. Are you planning on doing this before merging this PR?

If not, I wonder how much work is it? hours? days?

Reply:
The only reason to move to procedural macros is to get the correct source locations.

Reply:
Alright. Happy to tackle all three next once we figure out the syntax for the AST types.